utility/service/service_test.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/pokt-network/pocket/runtime/configs"
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+)
+
+func TestAdmitRelay(t *testing.T) {
+	testCases := []struct {
+		name          string
+		config        configs.ServicerConfig
+		relay         coreTypes.Relay
+		currentHeight int64
+		session       *coreTypes.Session
+		errSession    error
+		expected      error
+	}{
+		{
+			name:     "Relay with empty Meta is rejected",
+			expected: fmt.Errorf("Error admitting relay: relay metadata failed validation: empty relay metadata"),
+		},
+		{
+			name:     "Relay with unspecified chain is rejected",
+			relay:    coreTypes.Relay{Meta: &coreTypes.RelayMeta{}},
+			expected: fmt.Errorf("Error admitting relay: relay metadata failed validation: relay chain unspecified"),
+		},
+		{
+			name:     "Relay for unsupported chain is rejected",
+			relay:    testRelay("0021", 0),
+			expected: fmt.Errorf("Error admitting relay: relay metadata failed validation: relay chain not supported: %s", "0021"),
+		},
+		{
+			name:     "Relay with height set in a past session is rejected",
+			relay:    testRelay("0021", 5),
+			config:   testServicerConfig("0021"),
+			session:  testSession(sessionNumber(2), sessionBlocks(4), sessionHeight(2)),
+			expected: fmt.Errorf("Error admitting relay: relay failed block height validation: relay block height 5 not within session ID session-1 starting block 8 and last block 10"),
+		},
+		{
+			name:     "Relay with height set in a future session is rejected",
+			relay:    testRelay("0021", 9999),
+			config:   testServicerConfig("0021"),
+			session:  testSession(sessionNumber(2), sessionBlocks(4), sessionHeight(2)),
+			expected: fmt.Errorf("Error admitting relay: relay failed block height validation: relay block height 9999 not within session ID session-1 starting block 8 and last block 10"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			servicer := servicer{
+				provider: mockProvider{
+					Height:     tc.currentHeight,
+					Session:    tc.session,
+					ErrSession: tc.errSession,
+				},
+				config: tc.config,
+			}
+
+			err := servicer.admitRelay(tc.relay)
+			switch {
+			case err == nil && tc.expected != nil:
+				t.Fatalf("Expected error %v, got: nil", tc.expected)
+			case err != nil && tc.expected == nil:
+				t.Fatalf("Unexpected error %v", err)
+			case err != nil && err.Error() != tc.expected.Error():
+				t.Fatalf("Expected error %v, got: %v", tc.expected, err)
+			}
+		})
+	}
+}
+
+type mockProvider struct {
```

Comment:
We use [mockgen](https://github.com/golang/mock) throughout the entire codebase. If you search for `EXPECT()`, you should find a few examples as a starting point and I belive it's important to keep using that pattern throughout.

Lmk if you need more direct pointers or examples.

Reply:
Thank you for the review. Fixed.